### ISSUE TITLE
Send Buffer Access

### ISSUE THREAD

**Issue by wesbland** [wesbland/mpi-forum-historic#45]

**Comment by wesbland** [comment]
**Originally by _erezh@microsoft.com_ on 2008-10-16 15:48:54 -0500**

___
## Comments
Comment by Dick Treumann - The suggested change for collective communication is to strong. CC operations have both send and receive buffers. Before return from a CC, any access (not just modify) to a CC receive buffer must remain invalid.  The argument for access before completion applies only to CC send buffers

Reply by Erez Haba - The suggested wording is __less__ restrictive and say nothing about accessing the buffers for read. I think that it is obvious that reading receive buffer before the call is complete, results in invalid data. I did not want to add wording to explicitly allow reading the send buffer.

**Comment by wesbland** [comment]
**Originally by _buntinas@mcs.anl.gov_ on 2008-12-05 16:03:52 -0600**

___
Darius will review.

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2008-12-05 16:11:46 -0600**

___
By the way of reviewing this proposal, let's consider the following scenario:
 
- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.
- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.
- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.
 
What will happen one tries to access the send buffer in between?

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2008-12-05 16:15:59 -0600**

___
Another scenario for review:

- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order
- The data is transferred in background
- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete.

What will happen if this buffer is accessed in between?

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2008-12-05 16:19:12 -0600**

___
Transforming two notes above into comments (still learning).

I don't think proposal is good for MPI implementations. It restricts the internal MPI working way too much.

Let's consider the following scenario:

- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.

- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.

- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.

What will happen one tries to access the send buffer in between? 

Another scenario for review: 

- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order

- The data is transferred in background

- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete. 

What will happen if this buffer is accessed in between?

**Comment by wesbland** [comment]
**Originally by _rlgraham@ornl.gov_ on 2008-12-13 17:23:23 -0600**

___
Reviewed as chapter author for chapter 3.  Changes ok - Rich Graham

**Comment by wesbland** [comment]
**Originally by _traff@par.univie.ac.at_ on 2008-12-14 19:39:15 -0600**

___
Textually ok (but I share some of the doubts about the contents of this proposal),
except for the collectives part, where I'd suggest either not to change the text
(except that "communication buffer", p.131, line 13, should be "communication buffers"), or extend the explanation.

Jesper

**Comment by wesbland** [comment]
**Originally by _rabenseifner@hlrs.de_ on 2008-12-15 18:39:50 -0600**

___
Please can you add at the beginning of the Change-Log:

Section 3.4 on page 38, Section 3.9 on page 68, and Section 5.1 on page 129.

**Comment by wesbland** [comment]
**Originally by _rabenseifner@hlrs.de_ on 2009-01-09 06:46:31 -0600**

___
Sorry, that I did not found earlier, i.e., before 2nd vote.
Please add the following modification of MPI-2.1:

-Section 3.7 Page 52 Lines 5-6*[[BR]]
**Change:**[[BR]]
  The sender should not ~~access~~ any part of the send buffer after a nonblocking
  send operation is called, until the send completes.

-*To:** *(change “access” to “modify”)*[[BR]]
  The sender should not __modify__ any part of the send buffer after a nonblocking
  send operation is called, until the send completes.

Add **Section 3.7.2, page 49,** to the Change-Log Entry.

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2009-04-03 04:31:08 -0500**

___
Replying to RolfRabenseifner:
> Sorry, that I did not found earlier, i.e., before 2nd vote.
> Please add the following modification of MPI-2.1:

Technically, this should be a separate ticket.

**Comment by wesbland** [comment]
**Originally by _moody20@llnl.gov_ on 2009-07-10 15:19:21 -0500**

___
The collectives chapter changes are complete and checked into svn for this ticket.  I've removed the pdf, since it may have been incomplete (I only implemented the collective chapter changes).  The page in the current pdf is MPI page 131 (absolute page 157 in the document) and the ChangeLog entry which shows up on page 560 (absolute 586).  These page numbers may change though if other tickets move text around.

**Comment by wesbland** [comment]
**Originally by _moody20@llnl.gov_ on 2009-07-10 15:29:11 -0500**

___
**Attachment added:** [`chap-coll.pdf`](https://github.com/wesbland/mpi-forum-historic/raw/master/attachments//45/chap-coll.pdf) (41.9 KiB)
Whoops, I guess I can't delete the file, so I've uploaded a new one to overwrite it -- this only shows the collectives chapter changes.

**Comment by wesbland** [comment]
**Originally by _erezh@microsoft.com_ on 2008-10-16 15:48:54 -0500**

___
## Comments
Comment by Dick Treumann - The suggested change for collective communication is to strong. CC operations have both send and receive buffers. Before return from a CC, any access (not just modify) to a CC receive buffer must remain invalid.  The argument for access before completion applies only to CC send buffers

Reply by Erez Haba - The suggested wording is __less__ restrictive and say nothing about accessing the buffers for read. I think that it is obvious that reading receive buffer before the call is complete, results in invalid data. I did not want to add wording to explicitly allow reading the send buffer.

**Comment by wesbland** [comment]
**Originally by _buntinas@mcs.anl.gov_ on 2008-12-05 16:03:52 -0600**

___
Darius will review.

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2008-12-05 16:11:46 -0600**

___
By the way of reviewing this proposal, let's consider the following scenario:
 
- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.
- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.
- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.
 
What will happen one tries to access the send buffer in between?

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2008-12-05 16:15:59 -0600**

___
Another scenario for review:

- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order
- The data is transferred in background
- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete.

What will happen if this buffer is accessed in between?

**Comment by wesbland** [comment]
**Originally by _alexander.supalov@intel.com_ on 2008-12-05 16:19:12 -0600**

___
Transforming two notes above into comments (still learning).

I don't think proposal is good for MPI implementations. It restricts the internal MPI working way too much.

Let's consider the following scenario:

- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.

- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.

- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.

What will happen one tries to access the send buffer in between? 

Another scenario for review: 

- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order

- The data is transferred in background

- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete. 

What will happen if this buffer is accessed in between?

**Comment by wesbland** [comment]
**Originally by _rlgraham@ornl.gov_ on 2008-12-13 17:23:23 -0600**

___
Reviewed as chapter author for chapter 3.  Changes ok - Rich Graham

**Comment by wesbland** [comment]
**Originally by _traff@par.univie.ac.at_ on 2008-12-14 19:39:15 -0600**

___
Textually ok (but I share some of the doubts about the contents of this proposal),
except for the collectives part, where I'd suggest either not to change the text
(except that "communication buffer", p.131, line 13, should be "communication buffers"), or extend the explanation.

Jesper

**Comment by wesbland** [comment]
**Originally by _rabenseifner@hlrs.de_ on 2008-12-15 18:39:50 -0600**

___
Please can you add at the beginning of the Change-Log:

Section 3.4 on page 38, Section 3.9 on page 68, and Section 5.1 on page 129.

**Comment by wesbland** [comment]
**Originally by _erezh AT microsoft DOT  COOM _ on 2008-10-16 15:48:54 -0500**

___
## Comments
Comment by Dick Treumann - The suggested change for collective communication is to strong. CC operations have both send and receive buffers. Before return from a CC, any access (not just modify) to a CC receive buffer must remain invalid.  The argument for access before completion applies only to CC send buffers

Reply by Erez Haba - The suggested wording is __less__ restrictive and say nothing about accessing the buffers for read. I think that it is obvious that reading receive buffer before the call is complete, results in invalid data. I did not want to add wording to explicitly allow reading the send buffer.

**Comment by wesbland** [comment]
**Originally by _buntinas AT mcs DOT anl DOT gov_ on 2008-12-05 16:03:52 -0600**

___
Darius will review.

**Comment by wesbland** [comment]
**Originally by _alexander DOT supalov AT intel DOT  COOM _ on 2008-12-05 16:11:46 -0600**

___
By the way of reviewing this proposal, let's consider the following scenario:
 
- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.
- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.
- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.
 
What will happen one tries to access the send buffer in between?

**Comment by wesbland** [comment]
**Originally by _alexander DOT supalov AT intel DOT  COOM _ on 2008-12-05 16:15:59 -0600**

___
Another scenario for review:

- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order
- The data is transferred in background
- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete.

What will happen if this buffer is accessed in between?

**Comment by wesbland** [comment]
**Originally by _alexander DOT supalov AT intel DOT  COOM _ on 2008-12-05 16:19:12 -0600**

___
Transforming two notes above into comments (still learning).

I don't think proposal is good for MPI implementations. It restricts the internal MPI working way too much.

Let's consider the following scenario:

- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.

- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.

- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.

What will happen one tries to access the send buffer in between? 

Another scenario for review: 

- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order

- The data is transferred in background

- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete. 

What will happen if this buffer is accessed in between?

**Comment by wesbland** [comment]
**Originally by _rlgraham AT ornl DOT gov_ on 2008-12-13 17:23:23 -0600**

___
Reviewed as chapter author for chapter 3.  Changes ok - Rich Graham

**Comment by wesbland** [comment]
**Originally by _traff AT par DOT univie DOT ac DOT at_ on 2008-12-14 19:39:15 -0600**

___
Textually ok (but I share some of the doubts about the contents of this proposal),
except for the collectives part, where I'd suggest either not to change the text
(except that "communication buffer", p.131, line 13, should be "communication buffers"), or extend the explanation.

Jesper

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2008-12-15 18:39:50 -0600**

___
Please can you add at the beginning of the Change-Log:

Section 3.4 on page 38, Section 3.9 on page 68, and Section 5.1 on page 129.

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2009-01-09 06:46:31 -0600**

___
Sorry, that I did not found earlier, i.e., before 2nd vote.
Please add the following modification of MPI-2.1:

-Section 3.7 Page 52 Lines 5-6*[[BR]]
**Change:**[[BR]]
  The sender should not ~~access~~ any part of the send buffer after a nonblocking
  send operation is called, until the send completes.

-*To:** *(change “access” to “modify”)*[[BR]]
  The sender should not __modify__ any part of the send buffer after a nonblocking
  send operation is called, until the send completes.

Add **Section 3.7.2, page 49,** to the Change-Log Entry.

**Comment by wesbland** [comment]
**Originally by _alexander DOT supalov AT intel DOT  COOM _ on 2009-04-03 04:31:08 -0500**

___
Replying to RolfRabenseifner:
> Sorry, that I did not found earlier, i.e., before 2nd vote.
> Please add the following modification of MPI-2.1:

Technically, this should be a separate ticket.

**Comment by wesbland** [comment]
**Originally by _moody20 AT llnl DOT gov_ on 2009-07-10 15:19:21 -0500**

___
The collectives chapter changes are complete and checked into svn for this ticket.  I've removed the pdf, since it may have been incomplete (I only implemented the collective chapter changes).  The page in the current pdf is MPI page 131 (absolute page 157 in the document) and the ChangeLog entry which shows up on page 560 (absolute 586).  These page numbers may change though if other tickets move text around.

**Comment by wesbland** [comment]
**Originally by _moody20 AT llnl DOT gov_ on 2009-07-10 15:29:11 -0500**

___
**Attachment added:** [`chap-coll.pdf`](https://github.com/wesbland/mpi-forum-historic/raw/master/attachments/45/chap-coll.pdf) (41.9 KiB)
Whoops, I guess I can't delete the file, so I've uploaded a new one to overwrite it -- this only shows the collectives chapter changes.

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2009-07-13 08:30:43 -0500**

___
**Attachment added:** [`change-log_ticket45_item5.pdf`](https://github.com/wesbland/mpi-forum-historic/raw/master/attachments/45/change-log_ticket45_item5.pdf) (85.0 KiB)

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2009-07-13 10:37:46 -0500**

___
PDF-Review of the part in Chap. 5 (coll): Okay

**Comment by wesbland** [comment]
**Originally by _traff AT par DOT univie DOT ac DOT at_ on 2009-07-13 10:44:37 -0500**

___
Change in one-sided chapter (removal of "rationale") has been carried out, reviewers
please check (full text of Chapter 11 attached)

**Comment by wesbland** [comment]
**Originally by _traff AT par DOT univie DOT ac DOT at_ on 2009-07-13 10:45:25 -0500**

___
**Attachment added:** [`onesidedtickets.pdf`](https://github.com/wesbland/mpi-forum-historic/raw/master/attachments/45/onesidedtickets.pdf) (312.3 KiB)

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2009-07-13 12:33:13 -0500**

___
PDF-Review of the part in Chap. 11 (rma): Okay

**Comment by wesbland** [comment]
**Originally by _rlgraham AT ornl DOT gov_ on 2009-08-09 21:10:36 -0500**

___
changes made to chapter 3

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2009-08-20 06:23:06 -0500**

___
**Attachment added:** [`ticket45_all.pdf`](https://github.com/wesbland/mpi-forum-historic/raw/master/attachments/45/ticket45_all.pdf) (504.9 KiB)

**Comment by wesbland** [comment]
**Originally by _rabenseifner AT hlrs DOT de_ on 2009-08-20 06:25:11 -0500**

___
All changes are done, therefore switched to "Waiting for PDF reviews".

First review: all okay.
